src/async_fuse/memfs/cache/memory_cache.rs

```diff
@@ -0,0 +1,439 @@
+    }
+
+    /// Get a block from the in-memory cache without fetch it from backend.
+    fn get_block_from_cache(&self, ino: INum, block_id: usize) -> Option<Block> {
+        let guard = pin();
+        let block = self
+            .map
+            .get(&ino, &guard)
+            .and_then(|file_cache| file_cache.read().get(&block_id).cloned());
+        block
+    }
+
+    /// Return true if success
+    fn merge_block_in_place(&self, ino: INum, block_id: usize, to_merge: &IoBlock) -> bool {
+        let guard = pin();
+
+        let res = if let Some(mut lock) = self.map.get(&ino, &guard).map(RwLock::write) {
+            if let Some(block) = lock.get_mut(&block_id) {
+                // TODO: Mark block as dirty
+                merge_two_blocks(block, to_merge);
+                true
+            } else {
+                false
+            }
+        } else {
+            false
+        };
+        res
+    }
+
+    /// Write a block into the cache without writing through.
+    ///
+    /// If an old block is to be evicted, returns it with its coordinate.
+    fn write_block_into_cache(
+        &self,
+        ino: INum,
+        block_id: usize,
+        block: Block,
+    ) -> Option<(BlockCoordinate, Block)>
+    where
+        P: EvictPolicy<BlockCoordinate>,
+    {
+        let to_be_evicted = self.policy.put(BlockCoordinate(ino, block_id));
+        let guard = pin();
+        // Try to get the evicted block
+        let evicted = to_be_evicted.and_then(|coord| {
+            self.map
+                .get(&coord.0, &guard)
+                .and_then(|file_cache| file_cache.write().remove(&coord.1))
+                .map(|evicted| (coord, evicted))
+        });
+        // Insert the written block into cache
+        self.map
+            .get_or_insert(ino, RwLock::new(StdHashMap::new()), &guard)
+            .write()
+            .insert(block_id, block);
+        evicted
+    }
+
+    /// Write a block into the cache and evict a block to backend (if needed).
+    async fn write_and_evict(&self, ino: INum, block_id: usize, block: Block)
+    where
+        P: EvictPolicy<BlockCoordinate>,
+        S: Storage,
+    {
+        let evicted = self.write_block_into_cache(ino, block_id, block);
+        if let Some((BlockCoordinate(e_ino, e_block), evicted)) = evicted {
+            self.backend
+                .store(e_ino, e_block, IoBlock::new(evicted, 0, self.block_size))
+                .await;
+        }
+    }
+}
+
+#[async_trait]
+impl<P, S> Storage for InMemoryCache<P, S>
+where
+    P: EvictPolicy<BlockCoordinate> + Send + Sync,
+    S: Storage + Send + Sync,
+{
+    /// Loads a block from the cache, or loads it from the backend.
+    ///
+    /// If backend returns a `None`, this will create a new block with zeros filled,
+    /// inserts it into the cache, and returns it to caller. That means, this method
+    /// will never return `None`.
+    async fn load(&self, ino: INum, block_id: usize) -> Option<Block> {
```

Comment:
I think the main logic of load and store is the same for both Memory and Disk.

For read:
1. read local cache.
    1. if local cache is missing, load the backend.
2. if there's a load function in the previous step, we may get a evicted block, handle it properly.
3. return actul data to the caller.

According to the previous steps, we can make a default implementation for the load function, and move the `load_from_local_cache` into the trait.

Reply:
The `InMemoryCache` implementation has been moved to #442.